{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goals\n",
    "\n",
    "* Check whether Chris' datasets are in the scBaseCount SQL database (STAR results)\n",
    "* Also created `prodC` on GCP for Chris' datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import plotnine as pn\n",
    "from pypika import Query, Table, functions as fn\n",
    "\n",
    "from SRAgent.db.connect import db_connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 4)\n",
    "pn.theme_set(pn.theme_bw())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/home/nickyoungblut/dev/python/scBaseCount_analysis')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get base of github repo\n",
    "base_dir = !git rev-parse --show-toplevel\n",
    "base_dir = Path(base_dir[0])\n",
    "base_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set tenant\n",
    "os.environ[\"DYNACONF\"] = \"prod\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>estimated_number_of_cells</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>brain</td>\n",
       "      <td>None</td>\n",
       "      <td>999f2a15-3d7e-440b-96ae-2c806799c08c</td>\n",
       "      <td>Harmonized single-cell landscape, intercellula...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>acute slice culture of glioma resection</td>\n",
       "      <td>None</td>\n",
       "      <td>999f2a15-3d7e-440b-96ae-2c806799c08c</td>\n",
       "      <td>Harmonized single-cell landscape, intercellula...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>tumor tissue</td>\n",
       "      <td>None</td>\n",
       "      <td>3f50314f-bdc9-40c6-8e4a-b0901ebfbe4c</td>\n",
       "      <td>Single-cell sequencing links multiregional imm...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>H1299 cells</td>\n",
       "      <td>None</td>\n",
       "      <td>d0e9c47b-4ce7-4f84-b182-eddcfa0b2658</td>\n",
       "      <td>Single-cell gene expression profiling of SARS-...</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2382 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     srx_accession feature estimated_number_of_cells lib_prep        tech_10x  \\\n",
       "0             None    None                      None    other  not_applicable   \n",
       "1             None    None                      None    other  not_applicable   \n",
       "...            ...     ...                       ...      ...             ...   \n",
       "2380          None    None                      None    other  not_applicable   \n",
       "2381          None    None                      None    other  not_applicable   \n",
       "\n",
       "          organism                                   tissue  \\\n",
       "0     Homo sapiens                                    brain   \n",
       "1     Homo sapiens  acute slice culture of glioma resection   \n",
       "...            ...                                      ...   \n",
       "2380  Homo sapiens                             tumor tissue   \n",
       "2381  Homo sapiens                              H1299 cells   \n",
       "\n",
       "     tissue_ontology_term_id                     czi_collection_id  \\\n",
       "0                       None  999f2a15-3d7e-440b-96ae-2c806799c08c   \n",
       "1                       None  999f2a15-3d7e-440b-96ae-2c806799c08c   \n",
       "...                      ...                                   ...   \n",
       "2380                    None  3f50314f-bdc9-40c6-8e4a-b0901ebfbe4c   \n",
       "2381                    None  d0e9c47b-4ce7-4f84-b182-eddcfa0b2658   \n",
       "\n",
       "                                    czi_collection_name  \\\n",
       "0     Harmonized single-cell landscape, intercellula...   \n",
       "1     Harmonized single-cell landscape, intercellula...   \n",
       "...                                                 ...   \n",
       "2380  Single-cell sequencing links multiregional imm...   \n",
       "2381  Single-cell gene expression profiling of SARS-...   \n",
       "\n",
       "                             notes  \n",
       "0     Metadata obtained by SRAgent  \n",
       "1     Metadata obtained by SRAgent  \n",
       "...                            ...  \n",
       "2380  Metadata obtained by SRAgent  \n",
       "2381  Metadata obtained by SRAgent  \n",
       "\n",
       "[2382 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_screcounter_star_results(\n",
    "    conn, \n",
    "    feature=\"GeneFull_Ex50pAS\"\n",
    "):\n",
    "    \"\"\"\n",
    "    Get the screcounter_star_results table\n",
    "    \n",
    "    Args:\n",
    "        conn: A database connection object (e.g., sqlite3, psycopg2, etc.)\n",
    "        feature: String of the feature to filter on\n",
    "    Returns:\n",
    "        A pandas DataFrame of records\n",
    "    \"\"\"\n",
    "    \n",
    "    # Define tables\n",
    "    meta_table = Table('srx_metadata')\n",
    "    star_table = Table('screcounter_star_results')\n",
    "    \n",
    "    # Build the star query with feature filter (case insensitive)\n",
    "    star_query = (\n",
    "        Query.from_(star_table)\n",
    "        .select('*')\n",
    "        .where(\n",
    "            fn.Lower(star_table.feature) == feature.lower()\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Build the main query with join\n",
    "    # Assuming you want all columns from both tables except feature from star and _at columns from meta\n",
    "    main_query = (\n",
    "        Query.from_(meta_table)\n",
    "        .left_join(star_table)\n",
    "        .on(meta_table.srx_accession == star_table.sample)\n",
    "        .select(\n",
    "            star_table.sample.as_('srx_accession'),\n",
    "            star_table.feature,\n",
    "            star_table.estimated_number_of_cells,\n",
    "            meta_table.lib_prep,\n",
    "            meta_table.tech_10x,\n",
    "            meta_table.organism,\n",
    "            meta_table.tissue,\n",
    "            meta_table.tissue_ontology_term_id,\n",
    "            meta_table.czi_collection_id,\n",
    "            meta_table.czi_collection_name,\n",
    "            meta_table.notes\n",
    "        )\n",
    "        .where(\n",
    "            meta_table.czi_collection_id != None and meta_table.czi_collection_id != \"\"\n",
    "        )\n",
    "        .where(\n",
    "            star_table.estimated_number_of_cells.isnull()\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    # Execute query and return as DataFrame\n",
    "    return pd.read_sql_query(str(main_query), conn)\n",
    "\n",
    "with db_connect() as conn:\n",
    "    df_cells = get_screcounter_star_results(conn, feature=\"GeneFull_Ex50pAS\")\n",
    "df_cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Location of Chris' dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/processed_datasets/scRecount/cellxgene/counted_SRXs')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = Path('/processed_datasets/scRecount/cellxgene/counted_SRXs/')\n",
    "data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>processing</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull_ExonOverIntron</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull_ExonOverIntron</td>\n",
       "      <td>raw</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22895</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>raw</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22896</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22897 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession                  feature processing  \\\n",
       "0       ERX8792024  GeneFull_ExonOverIntron   filtered   \n",
       "1       ERX8792024  GeneFull_ExonOverIntron        raw   \n",
       "...            ...                      ...        ...   \n",
       "22895  ERX11467033         GeneFull_Ex50pAS        raw   \n",
       "22896  ERX11467033         GeneFull_Ex50pAS   filtered   \n",
       "\n",
       "                                                    path  \n",
       "0      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "1      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "...                                                  ...  \n",
       "22895  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "22896  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "\n",
       "[22897 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find all matrix files\n",
    "matrix_info = []\n",
    "for f in data_dir.rglob(\"matrix.mtx.gz\"):\n",
    "    accession = f.parts[5]\n",
    "    feature = f.parts[7]\n",
    "    processing = f.parent.name\n",
    "    matrix_info.append([accession, feature, processing, f])\n",
    "\n",
    "matrix_info = pd.DataFrame(matrix_info, columns=[\"accession\", \"feature\", \"processing\", \"path\"])\n",
    "matrix_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>processing</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERX12558782</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22868</th>\n",
       "      <td>ERX9138882</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22896</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>filtered</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2034 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession           feature processing  \\\n",
       "4       ERX8792024  GeneFull_Ex50pAS   filtered   \n",
       "10     ERX12558782  GeneFull_Ex50pAS   filtered   \n",
       "...            ...               ...        ...   \n",
       "22868   ERX9138882  GeneFull_Ex50pAS   filtered   \n",
       "22896  ERX11467033  GeneFull_Ex50pAS   filtered   \n",
       "\n",
       "                                                    path  \n",
       "4      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "10     /processed_datasets/scRecount/cellxgene/counte...  \n",
       "...                                                  ...  \n",
       "22868  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "22896  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "\n",
       "[2034 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# high priority matrix files\n",
    "matrix_info[(matrix_info[\"processing\"] == \"filtered\") & (matrix_info[\"feature\"] == \"GeneFull_Ex50pAS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull_ExonOverIntron</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14675</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>Gene</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14676</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14677 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession                  feature  \\\n",
       "0       ERX8792024  GeneFull_ExonOverIntron   \n",
       "1       ERX8792024                 GeneFull   \n",
       "...            ...                      ...   \n",
       "14675  ERX11467033                     Gene   \n",
       "14676  ERX11467033         GeneFull_Ex50pAS   \n",
       "\n",
       "                                                    path  \n",
       "0      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "1      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "...                                                  ...  \n",
       "14675  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "14676  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "\n",
       "[14677 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find all summary files\n",
    "sum_info = []\n",
    "for f in data_dir.rglob(\"Summary.csv\"):\n",
    "    feature = f.parent.stem\n",
    "    accession = f.parent.parent.parent.stem\n",
    "    sum_info.append([accession, feature, f])\n",
    "\n",
    "sum_info = pd.DataFrame(sum_info, columns=[\"accession\", \"feature\", \"path\"])\n",
    "sum_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull_ExonOverIntron</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX8792024</td>\n",
       "      <td>GeneFull</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14675</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>Gene</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14676</th>\n",
       "      <td>ERX11467033</td>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>/processed_datasets/scRecount/cellxgene/counte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10892 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession                  feature  \\\n",
       "0       ERX8792024  GeneFull_ExonOverIntron   \n",
       "1       ERX8792024                 GeneFull   \n",
       "...            ...                      ...   \n",
       "14675  ERX11467033                     Gene   \n",
       "14676  ERX11467033         GeneFull_Ex50pAS   \n",
       "\n",
       "                                                    path  \n",
       "0      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "1      /processed_datasets/scRecount/cellxgene/counte...  \n",
       "...                                                  ...  \n",
       "14675  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "14676  /processed_datasets/scRecount/cellxgene/counte...  \n",
       "\n",
       "[10892 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter sum_info to just those with filtered matrix\n",
    "sum_info_f = sum_info[sum_info[\"accession\"].isin(matrix_info[\"accession\"])]\n",
    "sum_info_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2179"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of unique SRXs\n",
    "sum_info_f[\"accession\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, feature, path]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find duplicates\n",
    "x = sum_info_f[sum_info_f[\"feature\"] == \"GeneFull_Ex50pAS\"]\n",
    "df_dup = x[x[\"accession\"].duplicated()].sort_values(\"accession\")\n",
    "\n",
    "pd.options.display.max_rows = 100\n",
    "pd.options.display.max_colwidth = None\n",
    "df_dup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset display\n",
    "pd.options.display.max_rows = 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter to those with metadata in the SQL database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx_accession</th>\n",
       "      <th>lib_prep</th>\n",
       "      <th>tech_10x</th>\n",
       "      <th>organism</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>czi_collection_id</th>\n",
       "      <th>czi_collection_name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX6640190</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>spleen</td>\n",
       "      <td>None</td>\n",
       "      <td>38833785-fac5-48fd-944a-0f62a4c23ed1</td>\n",
       "      <td>Construction of a human cell landscape at single-cell level</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX6640168</td>\n",
       "      <td>other</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>lung</td>\n",
       "      <td>None</td>\n",
       "      <td>38833785-fac5-48fd-944a-0f62a4c23ed1</td>\n",
       "      <td>Construction of a human cell landscape at single-cell level</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2170</th>\n",
       "      <td>SRX12677803</td>\n",
       "      <td>10x_Genomics</td>\n",
       "      <td>3_prime_gex</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>thymus</td>\n",
       "      <td>UBERON:0002370</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2171</th>\n",
       "      <td>SRX13549173</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>not_applicable</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>liver</td>\n",
       "      <td>UBERON:0002107</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Metadata obtained by SRAgent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2172 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     srx_accession        lib_prep        tech_10x      organism  tissue  \\\n",
       "0       SRX6640190           other  not_applicable  Homo sapiens  spleen   \n",
       "1       SRX6640168           other  not_applicable  Homo sapiens    lung   \n",
       "...            ...             ...             ...           ...     ...   \n",
       "2170   SRX12677803    10x_Genomics     3_prime_gex  Mus musculus  thymus   \n",
       "2171   SRX13549173  not_applicable  not_applicable  Mus musculus   liver   \n",
       "\n",
       "     tissue_ontology_term_id                     czi_collection_id  \\\n",
       "0                       None  38833785-fac5-48fd-944a-0f62a4c23ed1   \n",
       "1                       None  38833785-fac5-48fd-944a-0f62a4c23ed1   \n",
       "...                      ...                                   ...   \n",
       "2170          UBERON:0002370                                  None   \n",
       "2171          UBERON:0002107                                  None   \n",
       "\n",
       "                                              czi_collection_name  \\\n",
       "0     Construction of a human cell landscape at single-cell level   \n",
       "1     Construction of a human cell landscape at single-cell level   \n",
       "...                                                           ...   \n",
       "2170                                                         None   \n",
       "2171                                                         None   \n",
       "\n",
       "                             notes  \n",
       "0     Metadata obtained by SRAgent  \n",
       "1     Metadata obtained by SRAgent  \n",
       "...                            ...  \n",
       "2170  Metadata obtained by SRAgent  \n",
       "2171  Metadata obtained by SRAgent  \n",
       "\n",
       "[2172 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_srx = sum_info_f[\"accession\"].unique().tolist()\n",
    "\n",
    "meta_table = Table('srx_metadata')\n",
    "query = (\n",
    "    Query\n",
    "    .from_(meta_table)\n",
    "    .select(\n",
    "        meta_table.srx_accession,\n",
    "        meta_table.lib_prep,\n",
    "        meta_table.tech_10x,\n",
    "        meta_table.organism,\n",
    "        meta_table.tissue,\n",
    "        meta_table.tissue_ontology_term_id,\n",
    "        meta_table.czi_collection_id,\n",
    "        meta_table.czi_collection_name,\n",
    "        meta_table.notes\n",
    "    )\n",
    "    .where(\n",
    "        meta_table.srx_accession.isin(target_srx)\n",
    "    )\n",
    ")\n",
    "    \n",
    "with db_connect() as conn:\n",
    "    df_in_db = pd.read_sql_query(str(query), conn)\n",
    "df_in_db "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2172"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter chris's data to those in the SQL database\n",
    "x = df_in_db[\"srx_accession\"].unique().tolist()\n",
    "sum_info_ff = sum_info_f[sum_info_f[\"accession\"].isin(x)]\n",
    "sum_info_ff[\"accession\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Accessions with no in metadata table of SQL database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to those missing database\n",
    "x = df_in_db[\"srx_accession\"].unique().tolist()\n",
    "sum_info_f_mis = sum_info_f[~sum_info_f[\"accession\"].isin(x)]\n",
    "sum_info_f_mis[\"accession\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written to /home/nickyoungblut/dev/python/scBaseCount_analysis/data/chris_data/2025-07-03_srx-add-meta.csv\n"
     ]
    }
   ],
   "source": [
    "# write output\n",
    "outfile = base_dir /  'data' / 'chris_data' / '2025-07-03_srx-add-meta.csv'\n",
    "sum_info_f_mis.to_csv(outfile, index=False)\n",
    "print(f\"File written to {outfile}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note: the 7 are outliers that will not be added to the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter to just those lacking STAR results in the SQL database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERX11467030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERX11467031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>SRX9777398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>SRX9777399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>741 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample\n",
       "0    ERX11467030\n",
       "1    ERX11467031\n",
       "..           ...\n",
       "739   SRX9777398\n",
       "740   SRX9777399\n",
       "\n",
       "[741 rows x 1 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_srx = sum_info_ff[\"accession\"].unique().tolist()\n",
    "\n",
    "star_table = Table('screcounter_star_results')\n",
    "query = (\n",
    "    Query\n",
    "    .from_(star_table)\n",
    "    .select(\n",
    "        star_table.sample\n",
    "    )\n",
    "    .distinct()\n",
    "    .where(\n",
    "        star_table.sample.isin(target_srx)\n",
    "    )\n",
    ")\n",
    "    \n",
    "with db_connect() as conn:\n",
    "    df_has_star = pd.read_sql_query(str(query), conn)\n",
    "df_has_star"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_848383/2693290625.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1431"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter to just those lacking an SRA experiment accession; thus, should be added to the STAR results table\n",
    "x = df_has_star[\"sample\"].unique().tolist()\n",
    "sum_info_ff = sum_info_ff[~sum_info_f[\"accession\"].isin(x)]\n",
    "sum_info_ff[\"accession\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>feature</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, feature, path]\n",
       "Index: []"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for duplidates\n",
    "x = sum_info_ff.copy()\n",
    "x = x[x[\"feature\"] == \"GeneFull_Ex50pAS\"]\n",
    "x[x[\"accession\"].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>accession</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene</td>\n",
       "      <td>1431</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GeneFull</td>\n",
       "      <td>1431</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GeneFull_Ex50pAS</td>\n",
       "      <td>1431</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GeneFull_ExonOverIntron</td>\n",
       "      <td>1431</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Velocyto</td>\n",
       "      <td>1428</td>\n",
       "      <td>1428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   feature  accession  path\n",
       "0                     Gene       1431  1431\n",
       "1                 GeneFull       1431  1431\n",
       "2         GeneFull_Ex50pAS       1431  1431\n",
       "3  GeneFull_ExonOverIntron       1431  1431\n",
       "4                 Velocyto       1428  1428"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for all features per accession\n",
    "pd.options.display.max_rows = 10\n",
    "sum_info_ff.groupby(\"feature\").count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset display\n",
    "pd.options.display.max_rows = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File written to /home/nickyoungblut/dev/python/scBaseCount_analysis/data/chris_data/2025-07-03_srx-add-star.csv\n"
     ]
    }
   ],
   "source": [
    "# write output\n",
    "outfile = base_dir /  'data' / 'chris_data' / '2025-07-03_srx-add-star.csv'\n",
    "sum_info_ff.to_csv(outfile, index=False)\n",
    "print(f\"File written to {outfile}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add to SQL database & `prodC` GCP bucket location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "csv_file = base_dir / 'data' / 'chris_data' / '2025-07-03_srx-add-star.csv'\n",
    "print(csv_file.exists())\n",
    "\n",
    "exe = base_dir / 'scripts' / 'ChrisData2SQL-DB.py'\n",
    "print(exe.exists())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#SBATCH --job-name=ChrisData2SQLDB  \n",
      "#SBATCH --output=ChrisData2SQLDB.out\n",
      "#SBATCH --error=ChrisData2SQLDB.err\n",
      "#SBATCH --ntasks=1 \n",
      "#SBATCH --cpus-per-task=1\n",
      "#SBATCH --mem=8G\n",
      "#SBATCH --time=72:00:00\n",
      "\n",
      "/home/nickyoungblut/dev/python/scBaseCount_analysis/scripts/ChrisData2SQL-DB.py \\\n",
      "    --tenant test \\\n",
      "    --max-datasets 2 \\\n",
      "    --gcp-bucket gs://arc-ctc-screcounter/test/prodC/ \\\n",
      "    --samples-table /home/nickyoungblut/dev/python/scBaseCount_analysis/data/chris_data/2025-07-03_srx-add-star.csv\n"
     ]
    }
   ],
   "source": [
    "slurm_job = f\"\"\"#!/bin/bash\n",
    "#SBATCH --job-name=ChrisData2SQLDB  \n",
    "#SBATCH --output=ChrisData2SQLDB.out\n",
    "#SBATCH --error=ChrisData2SQLDB.err\n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=8G\n",
    "#SBATCH --time=72:00:00\n",
    "\n",
    "{exe} \\\\\n",
    "    --tenant test \\\\\n",
    "    --max-datasets 2 \\\\\n",
    "    --gcp-bucket gs://arc-ctc-screcounter/test/prodC/ \\\\\n",
    "    --samples-table {csv_file}\n",
    "\"\"\"\n",
    "\n",
    "job_file = \"ChrisData2SQL-DB.sh\"\n",
    "with open(job_file, \"w\") as f:\n",
    "    f.write(slurm_job)\n",
    "\n",
    "!chmod +x ChrisData2SQL-DB.sh\n",
    "!cat ChrisData2SQL-DB.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 934890\n"
     ]
    }
   ],
   "source": [
    "# submit job\n",
    "!sbatch ChrisData2SQL-DB.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-07-03 09:23:12,981 - Successfully uploaded matrix files for ERX12558782/Gene/raw\n",
      "2025-07-03 09:23:12,982 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12558782/ERR13187479_resultSolo.out/Gene/filtered -> gs://arc-ctc-screcounter/test/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12558782/Gene/filtered/\n",
      "2025-07-03 09:23:14,845 - Successfully uploaded matrix files for ERX12558782/Gene/filtered\n",
      "2025-07-03 09:23:14,846 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12558782/ERR13187479_resultSolo.out/GeneFull_ExonOverIntron/raw -> gs://arc-ctc-screcounter/test/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12558782/GeneFull_ExonOverIntron/raw/\n",
      "2025-07-03 09:23:16,812 - Successfully uploaded matrix files for ERX12558782/GeneFull_ExonOverIntron/raw\n",
      "2025-07-03 09:23:16,812 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12558782/ERR13187479_resultSolo.out/GeneFull_ExonOverIntron/filtered -> gs://arc-ctc-screcounter/test/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12558782/GeneFull_ExonOverIntron/filtered/\n",
      "2025-07-03 09:23:18,668 - Successfully uploaded matrix files for ERX12558782/GeneFull_ExonOverIntron/filtered\n",
      "2025-07-03 09:23:18,668 - Upload batch SCRECOUNTER_2025-07-03_00-00-00 completed\n",
      "2025-07-03 09:23:18,668 - Upload completed for batch: SCRECOUNTER_2025-07-03_00-00-00\n",
      "2025-07-03 09:23:18,668 - Updating screcounter_star_results...\n"
     ]
    }
   ],
   "source": [
    "!tail ChrisData2SQLDB.err"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Full run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "#SBATCH --job-name=ChrisData2SQLDB  \n",
      "#SBATCH --output=ChrisData2SQLDB.out\n",
      "#SBATCH --error=ChrisData2SQLDB.err\n",
      "#SBATCH --ntasks=1 \n",
      "#SBATCH --cpus-per-task=1\n",
      "#SBATCH --mem=12G\n",
      "#SBATCH --time=96:00:00\n",
      "\n",
      "/home/nickyoungblut/dev/python/scBaseCount_analysis/scripts/ChrisData2SQL-DB.py \\\n",
      "    --tenant prod \\\n",
      "    --ignore-missing \\\n",
      "    --gcp-bucket gs://arc-ctc-screcounter/prodC/ \\\n",
      "    --samples-table /home/nickyoungblut/dev/python/scBaseCount_analysis/data/chris_data/2025-07-03_srx-add-star.csv\n"
     ]
    }
   ],
   "source": [
    "slurm_job = f\"\"\"#!/bin/bash\n",
    "#SBATCH --job-name=ChrisData2SQLDB  \n",
    "#SBATCH --output=ChrisData2SQLDB.out\n",
    "#SBATCH --error=ChrisData2SQLDB.err\n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=96:00:00\n",
    "\n",
    "{exe} \\\\\n",
    "    --tenant prod \\\\\n",
    "    --ignore-missing \\\\\n",
    "    --gcp-bucket gs://arc-ctc-screcounter/prodC/ \\\\\n",
    "    --samples-table {csv_file}\n",
    "\"\"\"\n",
    "\n",
    "job_file = \"ChrisData2SQL-DB.sh\"\n",
    "with open(job_file, \"w\") as f:\n",
    "    f.write(slurm_job)\n",
    "\n",
    "!chmod +x ChrisData2SQL-DB.sh\n",
    "!cat ChrisData2SQL-DB.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 935000\n"
     ]
    }
   ],
   "source": [
    "# submit job\n",
    "!sbatch ChrisData2SQL-DB.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-07-03 13:09:09,871 - Successfully uploaded matrix files for ERX12098962/GeneFull_ExonOverIntron/raw\n",
      "2025-07-03 13:09:09,871 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12098962/ERR12724463_resultSolo.out/GeneFull_ExonOverIntron/filtered -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12098962/GeneFull_ExonOverIntron/filtered/\n",
      "2025-07-03 13:09:14,022 - Successfully uploaded matrix files for ERX12098962/GeneFull_ExonOverIntron/filtered\n",
      "2025-07-03 13:09:14,023 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12098962/ERR12724463_resultSolo.out/Gene/raw -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12098962/Gene/raw/\n",
      "2025-07-03 13:09:21,595 - Successfully uploaded matrix files for ERX12098962/Gene/raw\n",
      "2025-07-03 13:09:21,597 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12098962/ERR12724463_resultSolo.out/Gene/filtered -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12098962/Gene/filtered/\n",
      "2025-07-03 13:09:24,086 - Successfully uploaded matrix files for ERX12098962/Gene/filtered\n",
      "2025-07-03 13:09:24,086 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12098962/ERR12724463_resultSolo.out/GeneFull/raw -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12098962/GeneFull/raw/\n",
      "2025-07-03 13:09:30,502 - Successfully uploaded matrix files for ERX12098962/GeneFull/raw\n",
      "2025-07-03 13:09:30,503 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/ERX12098962/ERR12724463_resultSolo.out/GeneFull/filtered -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/ERX12098962/GeneFull/filtered/\n",
      "2025-07-03 13:09:33,336 - Successfully uploaded matrix files for ERX12098962/GeneFull/filtered\n",
      "2025-07-03 13:09:33,337 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/SRX22985533/SRR27308183_resultSolo.out/GeneFull_ExonOverIntron/raw -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/SRX22985533/GeneFull_ExonOverIntron/raw/\n",
      "2025-07-03 13:10:04,806 - Successfully uploaded matrix files for SRX22985533/GeneFull_ExonOverIntron/raw\n",
      "2025-07-03 13:10:04,811 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/SRX22985533/SRR27308183_resultSolo.out/GeneFull_ExonOverIntron/filtered -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/SRX22985533/GeneFull_ExonOverIntron/filtered/\n",
      "2025-07-03 13:10:13,760 - Successfully uploaded matrix files for SRX22985533/GeneFull_ExonOverIntron/filtered\n",
      "2025-07-03 13:10:13,761 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/SRX22985533/SRR27308183_resultSolo.out/Gene/raw -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/SRX22985533/Gene/raw/\n",
      "2025-07-03 13:10:21,059 - Successfully uploaded matrix files for SRX22985533/Gene/raw\n",
      "2025-07-03 13:10:21,059 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/SRX22985533/SRR27308183_resultSolo.out/Gene/filtered -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/SRX22985533/Gene/filtered/\n",
      "2025-07-03 13:10:24,928 - Successfully uploaded matrix files for SRX22985533/Gene/filtered\n",
      "2025-07-03 13:10:24,929 - Uploading matrix files: /processed_datasets/scRecount/cellxgene/counted_SRXs/SRX22985533/SRR27308183_resultSolo.out/GeneFull/raw -> gs://arc-ctc-screcounter/prodC/SCRECOUNTER_2025-07-03_00-00-00/STAR/SRX22985533/GeneFull/raw/\n"
     ]
    }
   ],
   "source": [
    "!tail -n 20 ChrisData2SQLDB.err"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"DYNACONF\"] = \"test\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  srx_accession database  entrez_id\n",
      "0    ERX8792024      sra   21270500\n"
     ]
    }
   ],
   "source": [
    "target = 'ERX8792024'\n",
    "\n",
    "\n",
    "srx_metadata = Table(\"srx_metadata\")\n",
    "stmt = Query \\\n",
    "    .from_(srx_metadata) \\\n",
    "    .select(\n",
    "        srx_metadata.srx_accession,\n",
    "        srx_metadata.database,\n",
    "        srx_metadata.entrez_id,\n",
    "    ) \\\n",
    "    .distinct() \\\n",
    "    .where(srx_metadata.srx_accession == target)\n",
    "        \n",
    "with db_connect() as conn:\n",
    "    print(pd.read_sql_query(str(stmt), conn))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /home/nickyoungblut/miniforge3/envs/SRAgent_nb:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "_python_abi3_support      1.0                  hd8ed1ab_2    conda-forge\n",
      "aiobotocore               2.23.0             pyhd8ed1ab_0    conda-forge\n",
      "aiohappyeyeballs          2.6.1              pyhd8ed1ab_0    conda-forge\n",
      "aiohttp                   3.11.14                  pypi_0    pypi\n",
      "aioitertools              0.12.0             pyhd8ed1ab_1    conda-forge\n",
      "aiosignal                 1.3.2              pyhd8ed1ab_0    conda-forge\n",
      "anndata                   0.11.4             pyhd8ed1ab_0    conda-forge\n",
      "annotated-types           0.7.0                    pypi_0    pypi\n",
      "anyio                     4.9.0                    pypi_0    pypi\n",
      "appdirs                   1.4.4                    pypi_0    pypi\n",
      "array-api-compat          1.12.0             pyhe01879c_0    conda-forge\n",
      "asgiref                   3.8.1                    pypi_0    pypi\n",
      "asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge\n",
      "attrs                     25.3.0             pyh71513ae_0    conda-forge\n",
      "aws-c-auth                0.8.7                h7743f02_1    conda-forge\n",
      "aws-c-cal                 0.8.7                h7d555fd_1    conda-forge\n",
      "aws-c-common              0.12.1               hb9d3cd8_0    conda-forge\n",
      "aws-c-compression         0.3.1                hcbd9e4e_3    conda-forge\n",
      "aws-c-event-stream        0.5.4                h286e7e7_3    conda-forge\n",
      "aws-c-http                0.9.5                hbca0721_0    conda-forge\n",
      "aws-c-io                  0.17.0               ha855f32_8    conda-forge\n",
      "aws-c-mqtt                0.12.2               hffac463_3    conda-forge\n",
      "aws-c-s3                  0.7.13               h4c9fe3b_3    conda-forge\n",
      "aws-c-sdkutils            0.2.3                hcbd9e4e_3    conda-forge\n",
      "aws-checksums             0.2.3                hcbd9e4e_3    conda-forge\n",
      "aws-crt-cpp               0.31.1               h46b750d_1    conda-forge\n",
      "aws-sdk-cpp               1.11.510             h1fa5cb7_4    conda-forge\n",
      "azure-core-cpp            1.14.0               h5cfcd09_0    conda-forge\n",
      "azure-identity-cpp        1.10.0               h113e628_0    conda-forge\n",
      "azure-storage-blobs-cpp   12.13.0              h3cf044e_1    conda-forge\n",
      "azure-storage-common-cpp  12.8.0               h736e048_1    conda-forge\n",
      "azure-storage-files-datalake-cpp 12.12.0              ha633028_1    conda-forge\n",
      "backoff                   2.2.1                    pypi_0    pypi\n",
      "bcrypt                    4.3.0                    pypi_0    pypi\n",
      "beautifulsoup4            4.13.3                   pypi_0    pypi\n",
      "biopython                 1.85                     pypi_0    pypi\n",
      "botocore                  1.38.27         pyge310_1234567_0    conda-forge\n",
      "brotli                    1.1.0                hb9d3cd8_2    conda-forge\n",
      "brotli-bin                1.1.0                hb9d3cd8_2    conda-forge\n",
      "brotli-python             1.1.0           py312h2ec8cdc_2    conda-forge\n",
      "build                     1.2.2.post1              pypi_0    pypi\n",
      "bzip2                     1.0.8                h4bc722e_7    conda-forge\n",
      "c-ares                    1.34.5               hb9d3cd8_0    conda-forge\n",
      "ca-certificates           2025.1.31            hbcca054_0    conda-forge\n",
      "cached-property           1.5.2                hd8ed1ab_1    conda-forge\n",
      "cached_property           1.5.2              pyha770c72_1    conda-forge\n",
      "cachetools                5.5.2                    pypi_0    pypi\n",
      "capnproto                 1.0.2                h766bdaa_3    conda-forge\n",
      "cellxgene-census          1.17.0             pyh29332c3_0    conda-forge\n",
      "certifi                   2025.1.31                pypi_0    pypi\n",
      "cffi                      1.17.1          py312h06ac9bb_0    conda-forge\n",
      "charset-normalizer        3.4.1                    pypi_0    pypi\n",
      "chroma-hnswlib            0.7.6                    pypi_0    pypi\n",
      "chromadb                  0.6.3                    pypi_0    pypi\n",
      "click                     8.1.8                    pypi_0    pypi\n",
      "colorama                  0.4.6              pyhd8ed1ab_1    conda-forge\n",
      "coloredlogs               15.0.1                   pypi_0    pypi\n",
      "comm                      0.2.2              pyhd8ed1ab_1    conda-forge\n",
      "contourpy                 1.3.1           py312h68727a3_0    conda-forge\n",
      "cpython                   3.12.11         py312hd8ed1ab_0    conda-forge\n",
      "cycler                    0.12.1             pyhd8ed1ab_1    conda-forge\n",
      "dataclasses-json          0.6.7                    pypi_0    pypi\n",
      "db-dtypes                 1.4.2                    pypi_0    pypi\n",
      "debugpy                   1.8.13          py312h2ec8cdc_0    conda-forge\n",
      "decorator                 5.2.1              pyhd8ed1ab_0    conda-forge\n",
      "deprecated                1.2.18                   pypi_0    pypi\n",
      "distro                    1.9.0                    pypi_0    pypi\n",
      "durationpy                0.9                      pypi_0    pypi\n",
      "dynaconf                  3.2.10                   pypi_0    pypi\n",
      "exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge\n",
      "executing                 2.1.0              pyhd8ed1ab_1    conda-forge\n",
      "fastapi                   0.115.11                 pypi_0    pypi\n",
      "filelock                  3.18.0                   pypi_0    pypi\n",
      "flatbuffers               25.2.10                  pypi_0    pypi\n",
      "fmt                       11.1.4               h07f6e7f_1    conda-forge\n",
      "fonttools                 4.56.0          py312h178313f_0    conda-forge\n",
      "freetype                  2.13.3               h48d6fc4_0    conda-forge\n",
      "frozenlist                1.5.0                    pypi_0    pypi\n",
      "fsspec                    2025.3.0                 pypi_0    pypi\n",
      "geos                      3.13.1               h97f6797_0    conda-forge\n",
      "gflags                    2.2.2             h5888daf_1005    conda-forge\n",
      "glog                      0.7.1                hbabe93e_0    conda-forge\n",
      "google-api-core           2.24.2                   pypi_0    pypi\n",
      "google-auth               2.38.0                   pypi_0    pypi\n",
      "google-cloud-bigquery     3.30.0                   pypi_0    pypi\n",
      "google-cloud-bigquery-storage 2.29.1                   pypi_0    pypi\n",
      "google-cloud-core         2.4.3                    pypi_0    pypi\n",
      "google-cloud-secret-manager 2.23.2                   pypi_0    pypi\n",
      "google-crc32c             1.7.0                    pypi_0    pypi\n",
      "google-resumable-media    2.7.2                    pypi_0    pypi\n",
      "googleapis-common-protos  1.69.2                   pypi_0    pypi\n",
      "greenlet                  3.1.1                    pypi_0    pypi\n",
      "grpc-google-iam-v1        0.14.2                   pypi_0    pypi\n",
      "grpcio                    1.71.0                   pypi_0    pypi\n",
      "grpcio-status             1.71.0                   pypi_0    pypi\n",
      "h11                       0.14.0                   pypi_0    pypi\n",
      "h2                        4.2.0              pyhd8ed1ab_0    conda-forge\n",
      "h5py                      3.13.0          nompi_py312hedeef09_100    conda-forge\n",
      "hdf5                      1.14.3          nompi_h2d575fe_109    conda-forge\n",
      "hpack                     4.1.0              pyhd8ed1ab_0    conda-forge\n",
      "httpcore                  1.0.7                    pypi_0    pypi\n",
      "httptools                 0.6.4                    pypi_0    pypi\n",
      "httpx                     0.28.1                   pypi_0    pypi\n",
      "httpx-sse                 0.4.0                    pypi_0    pypi\n",
      "huggingface-hub           0.29.3                   pypi_0    pypi\n",
      "humanfriendly             10.0                     pypi_0    pypi\n",
      "hyperframe                6.1.0              pyhd8ed1ab_0    conda-forge\n",
      "idna                      3.10               pyhd8ed1ab_1    conda-forge\n",
      "importlib-metadata        8.6.1              pyha770c72_0    conda-forge\n",
      "importlib-resources       6.5.2                    pypi_0    pypi\n",
      "iniconfig                 2.0.0                    pypi_0    pypi\n",
      "ipykernel                 6.29.5             pyh3099207_0    conda-forge\n",
      "ipython                   9.0.2              pyhfb0248b_0    conda-forge\n",
      "ipython_pygments_lexers   1.1.1              pyhd8ed1ab_0    conda-forge\n",
      "jedi                      0.19.2             pyhd8ed1ab_1    conda-forge\n",
      "jiter                     0.9.0                    pypi_0    pypi\n",
      "jmespath                  1.0.1              pyhd8ed1ab_1    conda-forge\n",
      "joblib                    1.5.1              pyhd8ed1ab_0    conda-forge\n",
      "jsonpatch                 1.33                     pypi_0    pypi\n",
      "jsonpointer               3.0.0                    pypi_0    pypi\n",
      "jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge\n",
      "jupyter_core              5.7.2              pyh31011fe_1    conda-forge\n",
      "keyutils                  1.6.1                h166bdaf_0    conda-forge\n",
      "kiwisolver                1.4.8           py312h84d6215_0    conda-forge\n",
      "krb5                      1.21.3               h659f571_0    conda-forge\n",
      "kubernetes                32.0.1                   pypi_0    pypi\n",
      "langchain                 0.3.20                   pypi_0    pypi\n",
      "langchain-chroma          0.2.2                    pypi_0    pypi\n",
      "langchain-community       0.3.19                   pypi_0    pypi\n",
      "langchain-core            0.3.45                   pypi_0    pypi\n",
      "langchain-openai          0.3.9                    pypi_0    pypi\n",
      "langchain-text-splitters  0.3.6                    pypi_0    pypi\n",
      "langchainhub              0.1.21                   pypi_0    pypi\n",
      "langgraph                 0.3.13                   pypi_0    pypi\n",
      "langgraph-checkpoint      2.0.21                   pypi_0    pypi\n",
      "langgraph-prebuilt        0.1.3                    pypi_0    pypi\n",
      "langgraph-sdk             0.1.57                   pypi_0    pypi\n",
      "langsmith                 0.3.15                   pypi_0    pypi\n",
      "lcms2                     2.17                 h717163a_0    conda-forge\n",
      "ld_impl_linux-64          2.43                 h712a8e2_4    conda-forge\n",
      "legacy-api-wrap           1.4.1              pyhd8ed1ab_0    conda-forge\n",
      "lerc                      4.0.0                h27087fc_0    conda-forge\n",
      "libabseil                 20250127.1      cxx17_hbbce691_0    conda-forge\n",
      "libaec                    1.1.4                h3f801dc_0    conda-forge\n",
      "libarrow                  19.0.1           hb90904d_7_cpu    conda-forge\n",
      "libarrow-acero            19.0.1           hcb10f89_7_cpu    conda-forge\n",
      "libarrow-dataset          19.0.1           hcb10f89_7_cpu    conda-forge\n",
      "libarrow-substrait        19.0.1           h1bed206_7_cpu    conda-forge\n",
      "libblas                   3.9.0           31_h59b9bed_openblas    conda-forge\n",
      "libbrotlicommon           1.1.0                hb9d3cd8_2    conda-forge\n",
      "libbrotlidec              1.1.0                hb9d3cd8_2    conda-forge\n",
      "libbrotlienc              1.1.0                hb9d3cd8_2    conda-forge\n",
      "libcblas                  3.9.0           31_he106b2a_openblas    conda-forge\n",
      "libcrc32c                 1.1.2                h9c3ff4c_0    conda-forge\n",
      "libcurl                   8.13.0               h332b0f4_0    conda-forge\n",
      "libdeflate                1.23                 h4ddbbb0_0    conda-forge\n",
      "libedit                   3.1.20250104    pl5321h7949ede_0    conda-forge\n",
      "libev                     4.33                 hd590300_2    conda-forge\n",
      "libevent                  2.1.12               hf998b51_1    conda-forge\n",
      "libexpat                  2.6.4                h5888daf_0    conda-forge\n",
      "libffi                    3.4.6                h2dba641_0    conda-forge\n",
      "libgcc                    14.2.0               h767d61c_2    conda-forge\n",
      "libgcc-ng                 14.2.0               h69a702a_2    conda-forge\n",
      "libgfortran               14.2.0               h69a702a_2    conda-forge\n",
      "libgfortran5              14.2.0               hf1ad2bd_2    conda-forge\n",
      "libgomp                   14.2.0               h767d61c_2    conda-forge\n",
      "libgoogle-cloud           2.36.0               hc4361e1_1    conda-forge\n",
      "libgoogle-cloud-storage   2.36.0               h0121fbd_1    conda-forge\n",
      "libgrpc                   1.71.0               he753a82_0    conda-forge\n",
      "libhwloc                  2.11.2          default_h0d58e46_1001    conda-forge\n",
      "libiconv                  1.18                 h4ce23a2_1    conda-forge\n",
      "libjpeg-turbo             3.0.0                hd590300_1    conda-forge\n",
      "liblapack                 3.9.0           31_h7ac8fdf_openblas    conda-forge\n",
      "liblzma                   5.6.4                hb9d3cd8_0    conda-forge\n",
      "libnghttp2                1.64.0               h161d5f1_0    conda-forge\n",
      "libnsl                    2.0.1                hd590300_0    conda-forge\n",
      "libopenblas               0.3.29          pthreads_h94d23a6_0    conda-forge\n",
      "libopentelemetry-cpp      1.20.0               hd1b1c89_0    conda-forge\n",
      "libopentelemetry-cpp-headers 1.20.0               ha770c72_0    conda-forge\n",
      "libparquet                19.0.1           h081d1f1_7_cpu    conda-forge\n",
      "libpng                    1.6.47               h943b412_0    conda-forge\n",
      "libprotobuf               5.29.3               h501fc15_1    conda-forge\n",
      "libre2-11                 2025.06.26           hba17884_0    conda-forge\n",
      "libsodium                 1.0.20               h4ab18f5_0    conda-forge\n",
      "libsqlite                 3.49.1               hee588c1_2    conda-forge\n",
      "libssh2                   1.11.1               hf672d98_0    conda-forge\n",
      "libstdcxx                 14.2.0               h8f9b012_2    conda-forge\n",
      "libstdcxx-ng              14.2.0               h4852527_2    conda-forge\n",
      "libthrift                 0.21.0               h0e7cc3e_0    conda-forge\n",
      "libtiff                   4.7.0                hd9ff511_3    conda-forge\n",
      "libtiledbsoma             1.16.2               h6b394ac_0    conda-forge\n",
      "libutf8proc               2.10.0               h202a827_0    conda-forge\n",
      "libuuid                   2.38.1               h0b41bf4_0    conda-forge\n",
      "libwebp-base              1.5.0                h851e524_0    conda-forge\n",
      "libxcb                    1.17.0               h8a09558_0    conda-forge\n",
      "libxcrypt                 4.4.36               hd590300_1    conda-forge\n",
      "libxml2                   2.13.7               h0d44e9d_0    conda-forge\n",
      "libzlib                   1.3.1                hb9d3cd8_2    conda-forge\n",
      "llvmlite                  0.44.0          py312h374181b_1    conda-forge\n",
      "lz4-c                     1.10.0               h5888daf_1    conda-forge\n",
      "markdown-it-py            3.0.0                    pypi_0    pypi\n",
      "marshmallow               3.26.1                   pypi_0    pypi\n",
      "matplotlib-base           3.10.1          py312hd3ec401_0    conda-forge\n",
      "matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge\n",
      "mdurl                     0.1.2                    pypi_0    pypi\n",
      "mizani                    0.13.1             pyhd8ed1ab_0    conda-forge\n",
      "mmh3                      5.1.0                    pypi_0    pypi\n",
      "monotonic                 1.6                      pypi_0    pypi\n",
      "more-itertools            10.7.0             pyhd8ed1ab_0    conda-forge\n",
      "mpmath                    1.3.0                    pypi_0    pypi\n",
      "msgpack                   1.1.0                    pypi_0    pypi\n",
      "multidict                 6.2.0                    pypi_0    pypi\n",
      "munkres                   1.1.4              pyh9f0ad1d_0    conda-forge\n",
      "mypy-extensions           1.0.0                    pypi_0    pypi\n",
      "natsort                   8.4.0              pyh29332c3_1    conda-forge\n",
      "ncurses                   6.5                  h2d0b736_3    conda-forge\n",
      "nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge\n",
      "networkx                  3.4.2                    pypi_0    pypi\n",
      "nlohmann_json             3.12.0               h3f2d84a_0    conda-forge\n",
      "numba                     0.61.2          py312h7bcfee6_1    conda-forge\n",
      "numpy                     1.26.4                   pypi_0    pypi\n",
      "oauthlib                  3.2.2                    pypi_0    pypi\n",
      "obonet                    1.1.0                    pypi_0    pypi\n",
      "onnxruntime               1.21.0                   pypi_0    pypi\n",
      "openai                    1.66.3                   pypi_0    pypi\n",
      "openjpeg                  2.5.3                h5fbd93e_0    conda-forge\n",
      "openssl                   3.4.1                h7b32b05_0    conda-forge\n",
      "opentelemetry-api         1.31.1                   pypi_0    pypi\n",
      "opentelemetry-exporter-otlp-proto-common 1.31.1                   pypi_0    pypi\n",
      "opentelemetry-exporter-otlp-proto-grpc 1.31.1                   pypi_0    pypi\n",
      "opentelemetry-instrumentation 0.52b1                   pypi_0    pypi\n",
      "opentelemetry-instrumentation-asgi 0.52b1                   pypi_0    pypi\n",
      "opentelemetry-instrumentation-fastapi 0.52b1                   pypi_0    pypi\n",
      "opentelemetry-proto       1.31.1                   pypi_0    pypi\n",
      "opentelemetry-sdk         1.31.1                   pypi_0    pypi\n",
      "opentelemetry-semantic-conventions 0.52b1                   pypi_0    pypi\n",
      "opentelemetry-util-http   0.52b1                   pypi_0    pypi\n",
      "orc                       2.1.1                h17f744e_1    conda-forge\n",
      "orjson                    3.10.15                  pypi_0    pypi\n",
      "overrides                 7.7.0                    pypi_0    pypi\n",
      "packaging                 24.2               pyhd8ed1ab_2    conda-forge\n",
      "pandas                    2.2.3           py312hf9745cd_1    conda-forge\n",
      "parso                     0.8.4              pyhd8ed1ab_1    conda-forge\n",
      "patsy                     1.0.1              pyhd8ed1ab_1    conda-forge\n",
      "pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge\n",
      "pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge\n",
      "pillow                    11.1.0          py312h80c1187_0    conda-forge\n",
      "pip                       25.0.1             pyh8b19718_0    conda-forge\n",
      "platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge\n",
      "plotnine                  0.14.5             pyhd8ed1ab_0    conda-forge\n",
      "pluggy                    1.5.0                    pypi_0    pypi\n",
      "polars                    1.31.0          default_h1650462_0    conda-forge\n",
      "polars-default            1.31.0           py39hfac2b71_0    conda-forge\n",
      "posthog                   3.21.0                   pypi_0    pypi\n",
      "prometheus-cpp            1.3.0                ha5d0236_0    conda-forge\n",
      "prompt-toolkit            3.0.50             pyha770c72_0    conda-forge\n",
      "propcache                 0.3.0                    pypi_0    pypi\n",
      "proto-plus                1.26.1                   pypi_0    pypi\n",
      "protobuf                  5.29.3                   pypi_0    pypi\n",
      "psutil                    7.0.0           py312h66e93f0_0    conda-forge\n",
      "psycopg2-binary           2.9.10                   pypi_0    pypi\n",
      "pthread-stubs             0.4               hb9d3cd8_1002    conda-forge\n",
      "ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge\n",
      "pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge\n",
      "pyarrow                   19.0.1                   pypi_0    pypi\n",
      "pyarrow-core              19.0.1          py312h01725c0_0_cpu    conda-forge\n",
      "pyarrow-hotfix            0.7                pyhd8ed1ab_0    conda-forge\n",
      "pyasn1                    0.6.1                    pypi_0    pypi\n",
      "pyasn1-modules            0.4.1                    pypi_0    pypi\n",
      "pycparser                 2.22               pyh29332c3_1    conda-forge\n",
      "pydantic                  2.10.6                   pypi_0    pypi\n",
      "pydantic-core             2.27.2                   pypi_0    pypi\n",
      "pydantic-settings         2.8.1                    pypi_0    pypi\n",
      "pygments                  2.19.1             pyhd8ed1ab_0    conda-forge\n",
      "pynndescent               0.5.13             pyhd8ed1ab_1    conda-forge\n",
      "pyparsing                 3.2.2              pyhd8ed1ab_0    conda-forge\n",
      "pypika                    0.48.9                   pypi_0    pypi\n",
      "pyproject-hooks           1.2.0                    pypi_0    pypi\n",
      "pysocks                   1.7.1              pyha55dd90_7    conda-forge\n",
      "pytest                    8.3.5                    pypi_0    pypi\n",
      "pytest-asyncio            0.25.3                   pypi_0    pypi\n",
      "python                    3.12.9          h9e4cc4f_1_cpython    conda-forge\n",
      "python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge\n",
      "python-dotenv             1.0.1                    pypi_0    pypi\n",
      "python-gil                3.12.11              hd8ed1ab_0    conda-forge\n",
      "python-tzdata             2025.2             pyhd8ed1ab_0    conda-forge\n",
      "python_abi                3.12                    5_cp312    conda-forge\n",
      "pytz                      2025.1                   pypi_0    pypi\n",
      "pyyaml                    6.0.2                    pypi_0    pypi\n",
      "pyzmq                     26.3.0          py312hbf22597_0    conda-forge\n",
      "qhull                     2020.2               h434a139_5    conda-forge\n",
      "re2                       2025.06.26           h9925aae_0    conda-forge\n",
      "readline                  8.2                  h8c095d6_2    conda-forge\n",
      "regex                     2024.11.6                pypi_0    pypi\n",
      "requests                  2.32.3                   pypi_0    pypi\n",
      "requests-oauthlib         2.0.0                    pypi_0    pypi\n",
      "requests-toolbelt         1.0.0                    pypi_0    pypi\n",
      "rich                      13.9.4                   pypi_0    pypi\n",
      "rsa                       4.9                      pypi_0    pypi\n",
      "s2n                       1.5.15               hd830067_0    conda-forge\n",
      "s3fs                      2025.5.1           pyhd8ed1ab_0    conda-forge\n",
      "scanpy                    1.11.2             pyhd8ed1ab_0    conda-forge\n",
      "scikit-learn              1.7.0           py312h7a48858_1    conda-forge\n",
      "scipy                     1.15.2          py312ha707e6e_0    conda-forge\n",
      "seaborn                   0.13.2               hd8ed1ab_3    conda-forge\n",
      "seaborn-base              0.13.2             pyhd8ed1ab_3    conda-forge\n",
      "session-info2             0.1.2              pyhd8ed1ab_0    conda-forge\n",
      "setuptools                75.8.2             pyhff2d567_0    conda-forge\n",
      "shapely                   2.1.1           py312h21f5128_0    conda-forge\n",
      "shellingham               1.5.4                    pypi_0    pypi\n",
      "six                       1.17.0             pyhd8ed1ab_0    conda-forge\n",
      "snappy                    1.2.1                h8bd8927_1    conda-forge\n",
      "sniffio                   1.3.1                    pypi_0    pypi\n",
      "somacore                  1.0.28             pyh29332c3_0    conda-forge\n",
      "soupsieve                 2.6                      pypi_0    pypi\n",
      "spdlog                    1.15.3               h10b92b3_0    conda-forge\n",
      "sqlalchemy                2.0.39                   pypi_0    pypi\n",
      "sragent                   0.1.1                    pypi_0    pypi\n",
      "stack_data                0.6.3              pyhd8ed1ab_1    conda-forge\n",
      "starlette                 0.46.1                   pypi_0    pypi\n",
      "statsmodels               0.14.4          py312hc0a28a1_0    conda-forge\n",
      "sympy                     1.13.3                   pypi_0    pypi\n",
      "tabulate                  0.9.0                    pypi_0    pypi\n",
      "tbb                       2022.1.0             h4ce085d_0    conda-forge\n",
      "tenacity                  9.0.0                    pypi_0    pypi\n",
      "threadpoolctl             3.6.0              pyhecae5ae_0    conda-forge\n",
      "tiktoken                  0.9.0                    pypi_0    pypi\n",
      "tiledb                    2.27.2               haba8ea6_5    conda-forge\n",
      "tiledb-py                 0.34.0          py312h9777f10_0    conda-forge\n",
      "tiledbsoma-py             1.16.2          py312h6095420_0    conda-forge\n",
      "tk                        8.6.13          noxft_h4845f30_101    conda-forge\n",
      "tokenizers                0.21.1                   pypi_0    pypi\n",
      "tornado                   6.4.2           py312h66e93f0_0    conda-forge\n",
      "tqdm                      4.67.1             pyhd8ed1ab_1    conda-forge\n",
      "traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge\n",
      "typer                     0.15.2                   pypi_0    pypi\n",
      "types-requests            2.32.0.20250306          pypi_0    pypi\n",
      "typing-extensions         4.12.2               hd8ed1ab_1    conda-forge\n",
      "typing-inspect            0.9.0                    pypi_0    pypi\n",
      "typing_extensions         4.12.2             pyha770c72_1    conda-forge\n",
      "tzdata                    2025.1                   pypi_0    pypi\n",
      "umap-learn                0.5.7           py312h7900ff3_1    conda-forge\n",
      "unicodedata2              16.0.0          py312h66e93f0_0    conda-forge\n",
      "urllib3                   2.3.0                    pypi_0    pypi\n",
      "uvicorn                   0.34.0                   pypi_0    pypi\n",
      "uvloop                    0.21.0                   pypi_0    pypi\n",
      "watchfiles                1.0.4                    pypi_0    pypi\n",
      "wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge\n",
      "websocket-client          1.8.0                    pypi_0    pypi\n",
      "websockets                15.0.1                   pypi_0    pypi\n",
      "wheel                     0.45.1             pyhd8ed1ab_1    conda-forge\n",
      "wrapt                     1.17.2          py312h66e93f0_0    conda-forge\n",
      "xmltodict                 0.14.2                   pypi_0    pypi\n",
      "xorg-libxau               1.0.12               hb9d3cd8_0    conda-forge\n",
      "xorg-libxdmcp             1.1.5                hb9d3cd8_0    conda-forge\n",
      "yarl                      1.18.3                   pypi_0    pypi\n",
      "zeromq                    4.3.5                h3b0a872_7    conda-forge\n",
      "zipp                      3.21.0             pyhd8ed1ab_1    conda-forge\n",
      "zlib                      1.3.1                hb9d3cd8_2    conda-forge\n",
      "zstandard                 0.23.0                   pypi_0    pypi\n",
      "zstd                      1.5.7                hb8e6e7a_2    conda-forge\n"
     ]
    }
   ],
   "source": [
    "!conda list"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SRAgent_nb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
